### PR TITLE
Implement other binary and unary ops in evaluator

### PR DESCRIPTION
**Pull request by alancai98**
Implements most of the other binary ops (`AND`, `OR`, `=`, `!=`) and unary ops (`NOT`, unary `+`, unary `-`) in the evaluator. Also included in this PR:
- Renaming of `EvalBinop` -> `EvalBinOp`
- Refactoring of some of the existing expression testing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-value/src/lib.rs [hunk 1]
```diff
@@ -155,6 +155,34 @@ impl ops::Rem for Value {
     }
 }
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-value/src/lib.rs:157]: Gotcha. That'll definitely make the testing and code in `eval.rs` cleaner. I'll update in an upcoming commit.

[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-value/src/lib.rs:157]: Updated to use traits in 91a2dba5f48134c12a6af20373574a687547a98a.

File: partiql-value/src/lib.rs [hunk 2]
```diff
@@ -1014,4 +1132,74 @@ mod tests {
...
+        assert_eq!(Value::from(false), Value::Null.and(Value::from(false)));
+        assert_eq!(Value::from(false), Value::from(false).and(Value::Null));
+        assert_eq!(Value::from(false), Value::Missing.and(Value::from(false)));
+        assert_eq!(Value::from(false), Value::from(false).and(Value::Missing));
+
+        // Null propagation => Null
+        assert_eq!(Value::Null, Value::Null.and(Value::Null));
+        assert_eq!(Value::Null, Value::Missing.and(Value::Missing));
+        assert_eq!(Value::Null, Value::Null.and(Value::Missing));
+        assert_eq!(Value::Null, Value::Missing.and(Value::Null));
```
Comment (am357) [inline partiql-value/src/lib.rs:1170]:
```suggestion
        assert_eq!(Value::Null, Value::Missing.and(Value::Null));
        assert_eq!(Value::Null, Value::Null.and(Value::from(true)));
        assert_eq!(Value::Null, Value::Missing.and(Value::from(true)));
```

File: partiql-value/src/lib.rs [hunk 3]
```diff
@@ -1014,4 +1132,74 @@ mod tests {
...
+        assert_eq!(Value::from(true), Value::Null.or(Value::from(true)));
+        assert_eq!(Value::from(true), Value::from(true).or(Value::Null));
+        assert_eq!(Value::from(true), Value::Missing.or(Value::from(true)));
+        assert_eq!(Value::from(true), Value::from(true).or(Value::Missing));
+
+        // Null propagation => Null
+        assert_eq!(Value::Null, Value::Null.or(Value::Null));
+        assert_eq!(Value::Null, Value::Missing.or(Value::Missing));
+        assert_eq!(Value::Null, Value::Null.or(Value::Missing));
+        assert_eq!(Value::Null, Value::Missing.or(Value::Null));
```
Comment (am357) [inline partiql-value/src/lib.rs:1197]:
```suggestion
        assert_eq!(Value::Null, Value::Missing.or(Value::Null));
        assert_eq!(Value::Null, Value::Null.or(Value::from(false)));
        assert_eq!(Value::Null, Value::Missing.or(Value::from(false)));        
```

File: partiql-eval/src/eval.rs [hunk 4]
```diff
@@ -278,16 +278,42 @@ impl EvalExpr for EvalLitExpr {
...
+#[derive(Debug)]
+pub struct EvalBinOpExpr {
+    pub op: EvalBinOp,
     pub lhs: Box<dyn EvalExpr>,
     pub rhs: Box<dyn EvalExpr>,
 }
 
 // TODO we should replace this enum with some identifier that can be looked up in a symtab/funcregistry
 #[derive(Debug)]
 #[allow(dead_code)] // TODO remove once out of PoC
```
Comment (am357) [inline partiql-eval/src/eval.rs:295]: Can we remove this now?

**Review by am357 — COMMENTED** [review]
LGTM w/ minor suggestions and a question?

File: partiql-value/src/lib.rs (hunk 2 above)
Comment (alancai98) [reply partiql-value/src/lib.rs:1170]: Added

File: partiql-value/src/lib.rs (hunk 3 above)
Comment (alancai98) [reply partiql-value/src/lib.rs:1201]: Added

**Review by am357 — APPROVED** [review]

File: partiql-eval/src/eval.rs [hunk 5]
```diff
@@ -307,45 +331,52 @@ pub enum EvalBinop {
     Exp,
 }
 
-impl EvalExpr for EvalBinopExpr {
+impl EvalExpr for EvalBinOpExpr {
```
Comment (jpschorr) [inline partiql-eval/src/eval.rs:334]: Propagation and short-circuit logic non-localized here.

If we want `AND`/`OR` to short-circuit on booleans, and all other ops to short-circuit on null-propagation, something like the following may clean this up a bit:

```rust


impl EvalExpr for EvalBinopExpr {
    fn evaluate(&self, bindings: &Tuple, ctx: &dyn EvalContext) -> Value {
        #[inline]
        fn short_circuit(op: &EvalBinop, value: &Value) -> Option<Value> {
            match (op, value) {
                (EvalBinop::And, Value::Boolean(false)) => Some(false.into()),
                (EvalBinop::Or, Value::Boolean(true)) => Some(true.into()),
                (_, Value::Missing) => Some(Value::Missing),
                (_, Value::Null) => Some(Value::Null),
                _ => None,
            }
        }

        let lhs = self.lhs.evaluate(bindings, ctx);
        if let Some(propagate) = short_circuit(&self.op, &lhs) {
            return propagate;
        }

        let rhs = self.rhs.evaluate(bindings, ctx);
        match self.op {
            EvalBinop::And => lhs.and(rhs),
            EvalBinOp::Or => lhs.or(rhs),
            EvalBinOp::Concat => ...
            EvalBinOp::Eq => Value::from(lhs == rhs),
            EvalBinOp::Neq => Value::from(lhs != rhs),
            EvalBinOp::Gt => Boolean(lhs > rhs),
            EvalBinOp::Gteq => Boolean(lhs >= rhs),
            EvalBinOp::Lt => Boolean(lhs < rhs),
            EvalBinOp::Lteq => Boolean(lhs <= rhs),
            EvalBinOp::Add => lhs + rhs,
            EvalBinOp::Sub => lhs - rhs,
            EvalBinOp::Mul => lhs * rhs,
            EvalBinOp::Div => lhs / rhs,
            EvalBinOp::Mod => lhs % rhs,
            EvalBinOp::Exp => todo!("Exponentiation"),
        }
    }
}

```

Comment (alancai98) [reply partiql-eval/src/eval.rs:337]: Thanks for the refactoring suggestion. Think the approach you've outlined is cleaner. From our offline discussion, made some adjustments to also check the RHS for `MISSING` and `NULL` values for equality and comparison ops.

File: partiql-value/src/lib.rs [hunk 6]
```diff
@@ -302,6 +302,8 @@ impl NullableEq for Value {
 impl NullableOrd for Value {
     type Output = Self;
 
+    // TODO: comparison is not right for data type mismatches. Equality permits mistyped arguments
```
Comment (alancai98) [inline partiql-value/src/lib.rs:305]: Will address this in a future PR. Don't want to over-expand the scope for this PR

**Review by jpschorr — APPROVED** [review]

**Pull request merged**